dvc/remote/gdrive/__init__.py

```diff
@@ -0,0 +1,567 @@
+        key = (args, tuple(kwargs.items()))
+        # could do with just setdefault, but it would require
+        # create/delete a "default" Lock() object for each call, so it
+        # is better to lock a single one for a short time
+        with lock:
+            if key not in locks:
+                locks[key] = threading.Lock()
+        with locks[key]:
+            if key not in results:
+                results[key] = func(*args, **kwargs)
+        return results[key]
+
+    return wrapped
+
+
+class RemoteGDrive(RemoteBASE):
+    """Google Drive remote implementation
+
+    Example URLs:
+
+    Datasets/my-dataset inside "My Drive" folder:
+
+        gdrive://root/Datasets/my-dataset
+
+    Folder by ID (recommended):
+
+        gdrive://1r3UbnmS5B4-7YZPZmyqJuCxLVps1mASC
+
+        (get it https://drive.google.com/drive/folders/{here})
+
+    Dataset named "my-dataset" in the hidden application folder:
+
+        gdrive://dvc/my-dataset
+
+        (this one wouldn't be visible through Google Drive web UI and
+         couldn't be shared)
+    """
+
+    scheme = Schemes.GDRIVE
+    REGEX = r"^gdrive://.*$"
+    REQUIRES = {"google-auth-oauthlib": google_auth_oauthlib}
+    PARAM_CHECKSUM = "md5Checksum"
+    GOOGLEAPIS_BASE_URL = "https://www.googleapis.com/"
+    MIME_GOOGLE_APPS_FOLDER = "application/vnd.google-apps.folder"
+    SPACE_DRIVE = "drive"
+    SCOPE_DRIVE = "https://www.googleapis.com/auth/drive"
+    SPACE_APPDATA = "appDataFolder"
+    SCOPE_APPDATA = "https://www.googleapis.com/auth/drive.appdata"
+    TIMEOUT = (5, 60)
+
+    # Default credential is needed to show the string of "Data Version
+    # Control" in OAuth dialog application name and icon in authorized
+    # applications list in Google account security settings. Also, the
+    # quota usage is limited by the application defined by client_id.
+    # The good practice would be to suggest the user to create their
+    # own application credentials.
+    DEFAULT_CREDENTIALPATH = os.path.join(
+        os.path.dirname(__file__), "google-dvc-client-id.json"
+    )
+
+    def __init__(self, repo, config):
+
+        super(RemoteGDrive, self).__init__(repo, config)
+
+        self.url = config[Config.SECTION_REMOTE_URL].rstrip("/")
+
+        parsed = urlparse(self.url)
+
+        if parsed.netloc == "dvc":
+            default_scopes = self.SCOPE_APPDATA
+            self.space = self.SPACE_APPDATA
+        else:
+            default_scopes = self.SCOPE_DRIVE
+            self.space = self.SPACE_DRIVE
+
+        credentialpath = config.get(
+            Config.SECTION_GDRIVE_CREDENTIALPATH, self.DEFAULT_CREDENTIALPATH
+        )
+        scopes = config.get(
+            Config.SECTION_GDRIVE_SCOPES, default_scopes
+        ).split()
+        self.oauth2 = OAuth2(
+            credentialpath,
+            scopes,
+            self.repo.config.config[Config.SECTION_OAUTH2],
+        )
+
+        # RemoteGDrive uses the "prefix" folder as the base reference,
```

Comment:
But what if I want to have several remotes? E.g.
```
remote_base = drive://root/
remote_dir = remote://remote_base/dir
remote_external_cache = remote://remote_base/cache
```
?

Reply:
Hm, I see `gdrive://root/Datasets/my-dataset` in your example above. So `Datasets/my-dateset` is not a prefix?

Reply:
Ooops. I decided to do this because of performance reasons.

// This is a show stopper. In the private conversation with @efiop we decided to rewrite the prefix logic, and use the cache to fix performance issues.